### PR TITLE
[TFLite] Support GPU buffer binding with C API

### PR DESCRIPTION
**Pull request by asus4**
The binding model's Input/Output with MTLBuffer/OpenGL SSBO is considerably effective for performance. But it is only available with C++ API, it will be great if we can use this feature with C API and eventually, Swift and Java.

https://www.tensorflow.org/lite/performance/gpu_advanced#inputoutput_buffers_ios_c_api_only

This PR allows us to call `TFLGpuDelegateBindMetalBufferToTensor (Metal Delegate)` and `TfLiteGpuDelegateBindBufferToTensor (GL Delegate)` from the C API.

### CODE REVIEW

**Comment by abattery** [comment]
@impjdi could you review this PR?

**Review by impjdi — COMMENTED** [review]
I'm fine with you moving functions from metal_delegate_internal.h to metal_delegate.h.  we did that to hide away unused advanced functionalities, but if you want to revive, I guess it's okay.

However, I am not an owner of TFLite (I'm only in charge of TFLite GPU), and I'm not sure what to make of the other changes you bring to c_api_experimental.  I'll delegate that to someone else from TFLite, e.g. srjoglekar246 & jdduke

**Comment by srjoglekar246** [comment]
Hey @asus4 , can you provide some details how:

1. How you use these APIs in your code? (A code snippet of inference would be good)
2. What is the timeline for your project?

Internally, we are also looking into some of these buffer interop use-cases across all delegates, and we hope to land something by next quarter. Since these APIs are experimental I don't mind landing them sooner, I just want to ensure that your direction isn't totally different from ours.

**Comment by asus4** [comment]
Great to hear that you are planning this. I was testing to bind the OpenGL Buffer to the GPU Delegate V2 in [this branch](https://github.com/asus4/tensorflow/tree/bindbuffer-gpudelegate).

I've been using TFLite C-API from Unity and I was hoping to use this feature as a lot of the process in Unity is on the GPU. 

1.  [inference code snippet in Unity C#](https://github.com/asus4/tf-lite-unity-sample/blob/gpu-binding/Assets/Samples/GpuBind/GpuBindSample.cs)

2. There is no specific timeline in my project (AR + ML project), but I've [got issues regarding this issue.](https://github.com/asus4/tf-lite-unity-sample/issues/23)

**Comment by gbaned** [comment]
@asus4, @srjoglekar246  Any update on this PR? Please. Thanks!

File: tensorflow/lite/c/c_api_experimental.h [hunk 1]
```diff
@@ -146,6 +146,43 @@ TFL_CAPI_EXPORT extern void TfLiteInterpreterOptionsSetUseNNAPI(
...
+/// contain new nodes that replace 1 more nodes.
+/// 'delegate' must outlive the interpreter.
+/// Returns one of the following three status codes:
+/// 1. kTfLiteOk: Success.
+/// 2. kTfLiteDelegateError: Delegation failed due to an error in the
+/// delegate. The Interpreter has been restored to its pre-delegation state.
+/// NOTE: This undoes all delegates previously applied to the Interpreter.
+/// 3. kTfLiteError: Unexpected/runtime failure.
+/// WARNING: This is an experimental API and subject to change.
+TFL_CAPI_EXPORT extern TfLiteStatus TfLiteInterpreterModifyGraphWithDelegate(
```
Comment (srjoglekar246) [inline tensorflow/lite/c/c_api_experimental.h:171]: Why is this needed? Does `TfLiteInterpreterOptionsAddDelegate` from `c_api.h` not work? (I want to minimize the new things we add to support this)

File: tensorflow/lite/c/c_api_experimental.h [hunk 2]
```diff
@@ -146,6 +146,43 @@ TFL_CAPI_EXPORT extern void TfLiteInterpreterOptionsSetUseNNAPI(
...
+    const TfLiteInterpreter* interpreter, TfLiteDelegate* delegate);
+
+/// Returns the tensor index with the input index.
+///
+/// WARNING: This is an experimental API and subject to change.
+TFL_CAPI_EXPORT extern int32_t TfLiteInterpreterGetInputTensorIndex(
+    const TfLiteInterpreter* interpreter, int32_t input_index);
+
+/// Returns the tensor index with the output index.
```
Comment (srjoglekar246) [inline tensorflow/lite/c/c_api_experimental.h:180]: nit: Please reframe as `Returns the tensor index corresponding to the output tensor.` Same with input.

File: tensorflow/lite/ios/BUILD.apple [hunk 3]
```diff
@@ -84,6 +84,7 @@ strip_common_include_path_prefix(
     name = "strip_common_include_path_core",
     hdr_labels = [
         "//tensorflow/lite/c:c_api.h",
+        "//tensorflow/lite/c:c_api_experimental.h",
```
Comment (srjoglekar246) [inline tensorflow/lite/ios/BUILD.apple:87]: @yyoon Is this fine? or should there be a different BUILD target with this experimental header (I am not very familiar with iOS BUILDs)

File: tensorflow/lite/delegates/gpu/metal_delegate.h [hunk 4]
```diff
@@ -67,6 +68,17 @@ TFL_CAPI_EXPORT extern TfLiteDelegate* TFLGpuDelegateCreate(
 // Destroys a delegate created with `TFLGpuDelegateCreate` call.
 TFL_CAPI_EXPORT extern void TFLGpuDelegateDelete(TfLiteDelegate* delegate);
 
+// Binds Metal buffer to an input or an output tensor in the initialized
+// delegate. Bound buffer should have sufficient storage to accommodate all
+// elements of a tensor. For quantized model, the buffer is bound to internal
+// dequantized float32 tensor.
+// Returns non-zero on success, or zero otherwise.
+//
+// *** Must be called *after* `Interpreter::ModifyGraphWithDelegate`. ***
```
Comment (srjoglekar246) [inline tensorflow/lite/delegates/gpu/metal_delegate.h:77]: Please add a WARNING that this is experimental, so that its not taken to be a stable feature :-)

**Review by srjoglekar246 — CHANGES_REQUESTED** [review]
As mentioned earlier on this PR, I am okay with getting this in since its only adding experimental APIs on TFLite C & changing the delegate's own API.

I will let @impjdi review the metal delegate changes & @yyoon look at the iOS BUILD change. 

@asus4, can you add to this PR a snippet of how you will use this in your code? It will greatly help as a reference for this change in the future.

File: tensorflow/lite/delegates/gpu/metal_delegate.h (hunk 4 above)
Comment (yyoon) [reply tensorflow/lite/delegates/gpu/metal_delegate.h:77]: +1

File: tensorflow/lite/ios/BUILD.apple (hunk 3 above)
Comment (yyoon) [reply tensorflow/lite/ios/BUILD.apple:87]: Looks fine to me. We actually had a separate issue filed for adding C experimental API:
https://github.com/tensorflow/tensorflow/issues/46852

**Review by yyoon — COMMENTED** [review]
Looks good on the iOS build change side. I'll let @srjoglekar246 to give the final approval.

**Comment by gbaned** [comment]
@asus4  Can you please check @yyoon's, @srjoglekar246's comments and keep us posted ? Thanks!

File: tensorflow/lite/c/c_api_experimental.h (hunk 1 above)
Comment (asus4) [reply tensorflow/lite/c/c_api_experimental.h:171]: @srjoglekar246 
I exposed this method for `gl_delegate.h`. `TfLiteGpuDelegateBindBufferToTensor` needs to be called before `TfLiteInterpreterModifyGraphWithDelegate`.

https://github.com/tensorflow/tensorflow/blob/b1bd64aacd5e7a6a918cd8b8557dc9ce23e011bc/tensorflow/lite/delegates/gpu/gl_delegate.h#L115-L121

The order of process on gl_delegate:
 1. Make an interpreter to get model input/output's shape
 2. Create OpenGL SSBO and bind it with the GPU delegate
 3. Call `TfLiteInterpreterModifyGraphWithDelegate`

Comment (srjoglekar246) [reply tensorflow/lite/c/c_api_experimental.h:171]: Aah...you are right. Maybe add a note in the docstrings that users should prefer to use `TfLiteInterpreterOptionsAddDelegate` unless absolutely required (like for this use-case)

File: tensorflow/lite/delegates/gpu/metal_delegate.h (hunk 4 above)
Comment (asus4) [reply tensorflow/lite/delegates/gpu/metal_delegate.h:77]: Added in ef4ddc6700f4c84096b977aa963d99199a5cbbcb

File: tensorflow/lite/c/c_api_experimental.h (hunk 2 above)
Comment (asus4) [reply tensorflow/lite/c/c_api_experimental.h]: Fixed in aa7542243a8116a1468a81717ae3223610ef64ee

File: tensorflow/lite/c/c_api_experimental.h (hunk 1 above)
Comment (asus4) [reply tensorflow/lite/c/c_api_experimental.h:173]: Added a note in 8f99d6b85fe3e198db8ab60315a3db5aff68d8c7

**Comment by asus4** [comment]
@srjoglekar246 Here's a snippet of the output binding on Metal.

```mm
// Create the model and interpreter options.
TfLiteModel* model = TfLiteModelCreateFromFile("/path/to/model.tflite");
TfLiteInterpreterOptions* options = TfLiteInterpreterOptionsCreate();

// Create the metal delegate
TfLiteDelegate* delegate = TFLGpuDelegateCreate(TFLGpuDelegateOptionsDefault());
TfLiteInterpreterOptionsAddDelegate(options, delegate);
TfLiteInterpreter* interpreter = TfLiteInterpreterCreate(model, options);

// Allocate tensors and populate the input tensor data.
TfLiteInterpreterAllocateTensors(interpreter);
TfLiteTensor* input_tensor = TfLiteInterpreterGetInputTensor(interpreter, 0);
TfLiteTensorCopyFromBuffer(input_tensor, input.data(), input.size() * sizeof(float));

// Create metal buffer and bind to the output tensor[0].
int channels = 4; // Shold be multiple of 4 on Metal GPU buffer.
id<MTLBuffer> output_buffer = [self createMetalBufferWithWidth:width height:height channels:channels];
int output_tensor_index = TfLiteInterpreterGetOutputTensorIndex(interpreter, 0);
TFLGpuDelegateBindMetalBufferToTensor(delegate, output_tensor_index, output_buffer);

// Interpreter doesn't copy the output data into CPU memory.
TfLiteSetAllowBufferHandleOutput(interpreter, true);

TfLiteInterpreterInvoke(interpreter);

// Do post process in metal buffer witout copying output to CPU memory.
[self doPostProcess:output_buffer];
```

**Review by srjoglekar246 — APPROVED** [review]
Approving this. Code changes seem okay, and I checked with @impjdi that he is okay with the Metal delegate API change.
